File: CSX/Closure, Scope & Execution Context/saveOutput.js
// SAVEOUTPUT

// expected input: callbackFn, password
// expected output: return a function

// approach:
// store object in outer function
// nested function (string)
//   nested function has an if statement:
//   if string != password
//     obj[string] = callbackFn(string)
//     return callbackFn(string)
//   else if password == string
//     return obj
// return nested function

function saveOutput(cb, pass) {
  let obj = {};
  function saveThis(str) {
    if (str != pass) {
      obj[str] = cb(str);
      return cb(str);
    } else if (str == pass) {
      return obj
    }
  } return saveThis;
}

const multiplyBy2 = function(num) { return num * 2; };
const multBy2AndLog = saveOutput(multiplyBy2, 'boo');
console.log(multBy2AndLog(2)); 
console.log(multBy2AndLog(9)); 
console.log(multBy2AndLog('boo')); 

File: CSX - Python/Closure, Scope & Execution Context/save_output.py
def save_output(cb, password):
    cache = {}
    def inner(arg):
        nonlocal cache
        if arg == password:
            return cache
        cache[arg] = cb(arg)
        return cache[arg]
    return inner

def multiply_by_2(n):
    return n * 2

mult_by_2_and_log = save_output(multiply_by_2, 'boo')

print(mult_by_2_and_log(2))
print(mult_by_2_and_log(9))
print(mult_by_2_and_log('boo'))
